src/index.js

```diff
@@ -12,55 +12,145 @@
-    for (
-      let matches = regExp.exec(text);
-      matches !== null;
-      matches = regExp.exec(text)
-    ) {
-      const originalIndex = source.originalIndexFromIndex(matches.index);
-      const originalRange = [originalIndex, originalIndex + matches[0].length];
-      const excludes = matchPatterns(text, skipPatterns);
-      const doesContain = exclude =>
-        source.originalIndexFromIndex(exclude.startIndex) <= originalRange[0] &&
-        originalRange[1] <= source.originalIndexFromIndex(exclude.endIndex);
-      if (!excludes.some(doesContain) && !spell.correct(matches[0])) {
-        const suggestions = spell.suggest(matches[0]);
-        const fix =
-          suggestions.length === 1
-            ? fixer.replaceTextRange(originalRange, suggestions[0])
-            : undefined;
-        const message = `${matches[0]} -> ${suggestions.join(", ")}`;
-        const ruleError = new RuleError(message, {
-          index: originalIndex,
-          fix
-        });
-        report(node, ruleError);
+const constructNSpell = function (language) {
+  const base = require.resolve(`dictionary-${language}`);
+  const dictionary = fs.readFileSync(
+    path.join(base, "..", "index.dic"),
+    "utf-8"
+  );
+  const affix = fs.readFileSync(path.join(base, "..", "index.aff"), "utf-8");
+  return nspell(affix, dictionary);
+};
+
+const getRuleErrorBuilder = (nSpell, suggestCorrections, RuleError, fixer) => {
+  if (suggestCorrections) {
+    return (word, originalRange) => {
+      const suggestions = nSpell.suggest(word);
+      const fix =
+        suggestions.length === 1
+          ? fixer.replaceTextRange(originalRange, suggestions[0])
+          : undefined;
+      const message = `${word} -> ${suggestions.join(", ")}`;
+      return new RuleError(message, {
+        index: originalRange[0],
+        fix: fix,
+      });
+    };
+  } else {
+    return (word, originalRange) => {
+      return new RuleError(word, {
+        index: originalRange[0],
+      });
+    };
+  }
+};
+
+const doesExclusionOverlapWord = (
+  [exclusionRangeStart, exclusionRangeEnd],
+  [wordRangeStart, wordRangeEnd]
+) => {
+  const startsBeforeEndOfExclusion = wordRangeStart <= exclusionRangeEnd;
+  const endsAfterStartOfExclusion = wordRangeEnd >= exclusionRangeStart;
+  return startsBeforeEndOfExclusion && endsAfterStartOfExclusion;
+};
+
+const getNoExclusionsOverlapWord = (exclusions, indexOffset) => {
+  if (exclusions && exclusions.length > 0) {
+    return (originalRange) => {
+      return !exclusions.some(({ startIndex, endIndex }) =>
+        doesExclusionOverlapWord(
+          [startIndex + indexOffset, endIndex + indexOffset],
+          originalRange
+        )
+      );
+    };
+  }
+  return (originalRange) => true;
+};
+
+const reporter = function (
+  { Syntax, RuleError, fixer, report },
+  {
+    language = "en",
+    skipPatterns = [],
+    wordDefinitionRegex: optionWordDefinitionRegex = /[\w']+/g,
```

Comment:
- Did you mean that `wordDefinitionRegex` accepts not only `RegExp` object but also a string because the format `.textlintrc` is json that rejects `RegExp` notion? As you passed this option to `RegExp` constructor in following code, I guess so.

- In JavaScript, the abbreviation of regular expression is called `Regexp`, isn't it? I know some programming language, such as Rust, calls `regex` as the regular expression, but textlint is written in JavaScript. I prefer to use `wordDefinitionRegexp`. Could you tell me your opinion?

Reply:
1. Yes it accepts both, because from `.textlintrc` in json format it will be a string that comes in in `wordDefinitionRegex` parameter. However when you instantiate the rule configuration programmatically (like in the tests) you can give an actual `RegExp` as the `wordDefinitionRegex` parameter. I pass this to the `RegExp` constructor, because for each loop using `RegExp.exec` with the `g` flag set on the regular expression `RegExp` behaves as a stateful object [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec), so for every piece of text searched a new instance is required.
2. Since we accept both a `string` and a `RegExp` I didn't give much thought to what the best option was. `regex` seemed natural to me, but I will change this to `regexp`.